lightning/restore/restore.go

```diff
@@ -304,48 +305,209 @@ outside:
-			}
-			defer db.Close()
-			db.ExecContext(ctx, "SET SQL_MODE = ?", rc.cfg.TiDB.StrSQLMode)
-		}
+type schemaStmtType int
+
+const (
+	schemaCreateDatabase = iota
+	schemaCreateTable
+	schemaCreateView
+)
+
+type schemaJob struct {
+	dbName  string
+	session checkpoints.Session
+	stmts   []*schemaStmt
+}
+
+type schemaStmt struct {
+	tblName  string
+	stmtType schemaStmtType
+	sql      string
+}
 
-		for _, dbMeta := range rc.dbMetas {
-			task := log.With(zap.String("db", dbMeta.Name)).Begin(zap.InfoLevel, "restore table schema")
+type restoreSchemaWorker struct {
+	ctx      context.Context
+	quit     context.CancelFunc
+	jobCh    chan *schemaJob
+	errCh    chan error
+	wg       sync.WaitGroup
+	glue     glue.Glue
+	store    storage.ExternalStorage
+	sessions map[string]checkpoints.Session
+}
 
-			tablesSchema := make(map[string]string)
+func (worker *restoreSchemaWorker) makeJobs(dbMetas []*mydump.MDDatabaseMeta) {
+	//TODO: maybe we should put these create statements into a transaction
+	select {
+	case <-worker.ctx.Done():
+		return
+	default:
+		// 1. restore databases, execute statements concurrency
+		for _, dbMeta := range dbMetas {
+			restoreSchemaJob := &schemaJob{
+				dbName:  dbMeta.Name,
+				session: worker.getSession(dbMeta.Name),
+				stmts:   make([]*schemaStmt, 0, 1),
+			}
+			restoreSchemaJob.stmts = append(restoreSchemaJob.stmts, &schemaStmt{
+				stmtType: schemaCreateDatabase,
+				sql:      createDatabaseIfNotExistStmt(dbMeta.Name),
+			})
+			worker.wg.Add(1)
+			worker.jobCh <- restoreSchemaJob
+		}
+		worker.wg.Wait()
+		// 2. restore tables, execute statements concurrency
+		for _, dbMeta := range dbMetas {
 			for _, tblMeta := range dbMeta.Tables {
-				tablesSchema[tblMeta.Name] = tblMeta.GetSchema(ctx, rc.store)
+				sql := tblMeta.GetSchema(worker.ctx, worker.store)
+				if sql != "" {
+					stmts, err := createTableIfNotExistsStmt(worker.glue.GetParser(), sql, dbMeta.Name, tblMeta.Name)
+					if err != nil {
+						worker.throw(err)
+					}
+					for _, sql := range stmts {
+						restoreSchemaJob := &schemaJob{
+							dbName:  dbMeta.Name,
+							session: worker.getSession(dbMeta.Name),
+							stmts:   make([]*schemaStmt, 0, 1),
+						}
+						restoreSchemaJob.stmts = append(restoreSchemaJob.stmts, &schemaStmt{
+							tblName:  tblMeta.Name,
+							stmtType: schemaCreateTable,
+							sql:      sql,
+						})
+						worker.wg.Add(1)
+						worker.jobCh <- restoreSchemaJob
+					}
+				}
 			}
-			err := InitSchema(ctx, rc.tidbGlue, dbMeta.Name, tablesSchema)
-
-			task.End(zap.ErrorLevel, err)
-			if err != nil {
-				return errors.Annotatef(err, "restore table schema %s failed", dbMeta.Name)
+		}
+		worker.wg.Wait()
+		// 3. restore views. Since views can cross database we must restore views after all table schemas are restored.
+		for _, dbMeta := range dbMetas {
+			restoreSchemaJob := &schemaJob{
+				dbName:  dbMeta.Name,
+				session: worker.getSession(dbMeta.Name),
```

Comment:
should we use session per view not per db?

Reply:
It's better not, we execute view schema in serial, I think open more connections is not helpful. And we must keep the same session for every  restore schema job, avoid unable to read user variables declared earlier in the session.